src/main/java/com/getindata/connectors/http/internal/sink/httpclient/BatchRequestSubmitter.java

```diff
@@ -0,0 +1,140 @@
+package com.getindata.connectors.http.internal.sink.httpclient;
+
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpClient.Version;
+import java.net.http.HttpRequest.BodyPublisher;
+import java.net.http.HttpRequest.BodyPublishers;
+import java.net.http.HttpRequest.Builder;
+import java.net.http.HttpResponse;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Properties;
+import java.util.concurrent.CompletableFuture;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.flink.annotation.VisibleForTesting;
+
+import com.getindata.connectors.http.internal.config.HttpConnectorConfigConstants;
+import com.getindata.connectors.http.internal.sink.HttpSinkRequestEntry;
+
+/**
+ * This implementation groups received events in batches and submits each batch as individual HTTP
+ * requests. Batch is created based on batch size or based on HTTP method type.
+ */
+@Slf4j
+public class BatchRequestSubmitter extends AbstractRequestSubmitter {
+
+    private static final byte[] BATCH_START_BYTES = "[".getBytes(StandardCharsets.UTF_8);
+
+    private static final byte[] BATCH_END_BYTES = "]".getBytes(StandardCharsets.UTF_8);
+
+    private static final byte[] BATCH_ELEMENT_DELIM_BYTES = ",".getBytes(StandardCharsets.UTF_8);
+
+    private final int httpRequestBatchSize;
+
+    public BatchRequestSubmitter(
+            Properties properties,
+            String[] headersAndValue,
+            HttpClient httpClient) {
+
+        super(properties, headersAndValue, httpClient);
+
+        this.httpRequestBatchSize = Integer.parseInt(
+            properties.getProperty(HttpConnectorConfigConstants.SINK_HTTP_BATCH_REQUEST_SIZE)
+        );
+    }
+
+    @Override
+    public List<CompletableFuture<JavaNetHttpResponseWrapper>> submit(
+            String endpointUrl,
+            List<HttpSinkRequestEntry> requestsToSubmit) {
+
+        if (requestsToSubmit.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        var responseFutures = new ArrayList<CompletableFuture<JavaNetHttpResponseWrapper>>();
+        String previousReqeustMethod = requestsToSubmit.get(0).method;
+        List<HttpSinkRequestEntry> reqeustBatch = new ArrayList<>(httpRequestBatchSize);
+
+        for (var entry : requestsToSubmit) {
+            if (reqeustBatch.size() == httpRequestBatchSize
+                || !previousReqeustMethod.equalsIgnoreCase(entry.method)) {
+                // break batch and submit
+                responseFutures.add(sendBatch(endpointUrl, reqeustBatch));
+                reqeustBatch.clear();
+            }
+            reqeustBatch.add(entry);
+            previousReqeustMethod = entry.method;
+        }
+
+        // submit anything that left
+        responseFutures.add(sendBatch(endpointUrl, reqeustBatch));
```

Comment:
We might be sending empty requests here (or, looking at the rest of the code, an empty JSON list `[]`). I wonder whether it is expected or should we rather add `if (requestBatch.size() > 0) ...`?